### ISSUE TITLE
STDIN not working.

### ISSUE THREAD

**Issue by mikebronner** [phpmd/phpmd#882]
- PHPMD version: 2.10.0
- PHP Version: 8.0.3
- Installation type: phar file
- Operating System / Distribution & Version: macOS 11.4

### Current Behavior
According to issue #542 STDIN should be fixed upstream. Issue #185 is quite old, is that still an issue? I was hoping for STDIN functionality on this release, so I could implement it easily in am IDE plugin.

### Expected Behavior
Was hoping to be able to use STDIN for development of a plugin.

### Steps To Reproduce:
Run `phpmd - json cleancode,codesize,controversial,design,naming,unusedcode` does not lint anything.

## Checks before submitting
* [X] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [X] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [X] I have added every step to reproduce the bug.
* [X] If possible I added relevant code examples.
* [X] This issue is about 1 bug and nothing more.
* [X] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by kylekatarnls** [comment]
Hello, 👋 

STDIN implies you first output an other command as shown on https://github.com/phpmd/phpmd/issues/542#issuecomment-570605989:

I use `cat` with this simple file **test.php** (but it would work with any command so does not need any file):
```php
<?php

class Foo
{
    public function bar()
    {
        $a = 9;
    }
}
```
And I run:
```
cat test.php | phpmd - json naming
```
And get:
```json
{
    "version": "@package_version@",
    "package": "phpmd",
    "timestamp": "2021-04-28T07:53:22+00:00",
    "files": [
        {
            "file": "php:\/\/stdin",
            "violations": [
                {
                    "beginLine": 7,
                    "endLine": 7,
                    "package": null,
                    "function": null,
                    "class": null,
                    "method": null,
                    "description": "Avoid variables with short names like $a. Configured minimum length is 3.",
                    "rule": "ShortVariable",
                    "ruleSet": "Naming Rules",
                    "externalInfoUrl": "https:\/\/phpmd.org\/rules\/naming.html#shortvariable",
                    "priority": 3
                }
            ]
        }
    ]
}
```

So I can see STDIN is working as expected.

Please provide full steps to reproduce the issue if you still have troubles and I'll re-open.

Thanks.

**Comment by Geraint** [comment]
@kylekatarnls I'm getting some very confusing results with this, too.

I copied your `test.php` file from above, and ran it with this `phpmd.xml` file

```xml
<?xml version="1.0"?>
<ruleset name="Project Code Standards"
         xmlns="http://pmd.sf.net/ruleset/1.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://pmd.sf.net/ruleset/1.0.0
                     http://pmd.sf.net/ruleset_xml_schema.xsd"
         xsi:noNamespaceSchemaLocation="
                     http://pmd.sf.net/ruleset_xml_schema.xsd">
    <description>
        Rules to apply (or not apply) for said project
    </description>

    <rule ref="rulesets/cleancode.xml">
    </rule>

    <rule ref="rulesets/codesize.xml">
    </rule>

    <rule ref="rulesets/controversial.xml">
    </rule>

    <rule ref="rulesets/design.xml">
    </rule>

    <rule ref="rulesets/naming.xml">
    </rule>

    <rule ref="rulesets/unusedcode.xml">
    </rule>
</ruleset>
```

When I ran phpmd normally (i.e. not trying to use STDIN), I got these results:

```
./vendor/bin/phpmd test.php text phpmd.xml
/home/geraint/workspace/sandbox/phpmd-test/test.php:7  ShortVariable        Avoid variables with short names like $a. Configured minimum length is 3.
/home/geraint/workspace/sandbox/phpmd-test/test.php:7  UnusedLocalVariable  Avoid unused local variables such as '$a'.
```

However, when I ran phpmd using STDIN, I got some very strange results indeed:

```
cat test.php | ./vendor/bin/phpmd - text phpmd.xml
php://stdin:50   LongVariable    Avoid excessively long variable names like $redactedVariableName1. Keep variable name length under 20.
php://stdin:57   LongVariable    Avoid excessively long variable names like $redactedVariableName2. Keep variable name length under 20.
php://stdin:156  ShortVariable   Avoid variables with short names like $e. Configured minimum length is 3.
php://stdin:169  LongVariable    Avoid excessively long variable names like $redactedVariableName3. Keep variable name length under 20.
php://stdin:182  ElseExpression  The method redactedMethodName1 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
php://stdin:224  ElseExpression  The method redactedMethodName2 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
php://stdin:483  ElseExpression  The method redactedMethodName3 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
```

(Note: I have edited the variable and method names in the output)

This is strange, because that input is from a much bigger file that is in an entirely separate project.

I *had* been running phpmd against that other file, in my efforts to get phpmd to work with [null-ls.nvim](https://github.com/jose-elias-alvarez/null-ls.nvim) - it seems to be weirdly cached.

I got the same results, even when I did not pipe *anything* into phpmd:

```
./vendor/bin/phpmd - text phpmd.xml
php://stdin:50   LongVariable    Avoid excessively long variable names like $redactedVariableName1. Keep variable name length under 20.
php://stdin:57   LongVariable    Avoid excessively long variable names like $redactedVariableName2. Keep variable name length under 20.
php://stdin:156  ShortVariable   Avoid variables with short names like $e. Configured minimum length is 3.
php://stdin:169  LongVariable    Avoid excessively long variable names like $redactedVariableName3. Keep variable name length under 20.
php://stdin:182  ElseExpression  The method redactedMethodName1 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
php://stdin:224  ElseExpression  The method redactedMethodName2 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
php://stdin:483  ElseExpression  The method redactedMethodName3 uses an else expression. Else clauses are basically not necessary and you can simplify the code by not using them.
```

After some head scratching, I narrowed the problem down to a folder called `~/.pdepend`.

If I `rm -rf` it, things start behaving sanely.

To reproduce the problem, take your `test.php` file, and run the following:

```
cat test.php | ./vendor/bin/phpmd - text phpmd.xml
php://stdin:7  ShortVariable        Avoid variables with short names like $a. Configured minimum length is 3.
php://stdin:7  UnusedLocalVariable  Avoid unused local variables such as '$a'.
```

Now, edit the file, and change `$a` to `$b`, and run phpmd again:

```
cat test.php | ./vendor/bin/phpmd - text phpmd.xml
php://stdin:7  ShortVariable        Avoid variables with short names like $a. Configured minimum length is 3.
php://stdin:7  UnusedLocalVariable  Avoid unused local variables such as '$a'.
```

Note that phpmd is still reporting `$a` as the problem, even though that variable no longer exists.

Now, delete the `.pdepend` folder, and run phpmd again:

```
rm -rf ~/.pdepend
cat test.php | ./vendor/bin/phpmd - text phpmd.xml
php://stdin:7  ShortVariable        Avoid variables with short names like $b. Configured minimum length is 3.
php://stdin:7  UnusedLocalVariable  Avoid unused local variables such as '$b'.
```

phpmd is now correctly reporting a problem with `$b` and not `$a`.

This problem doesn't occur if I use a filename, rather than STDIN.

Unless this is somehow , it seems like STDIN is not working as expected, and this bug should still be open.

FWIW, I've seen this problem this using:

- PHP 7.4.27, 8.0.16, and 8.1.1
- PHPMD 2.11.1
- Running on Windows 10 / WSL2 / Ubuntu 20.04
